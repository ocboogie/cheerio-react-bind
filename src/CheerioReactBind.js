import React from "react";
import PropTypes from "prop-types";

export default class CheerioReactBind extends React.Component {
  constructor(props) {
    super(props);
    this.props.$elem.data("update", this.update.bind(this));
    this.update(true);
  }

  update(first = false) {
    const children = [];
    const $children = this.props.$elem.children();
    if ($children.length > 0) {
      this.props.$elem.children().each((index, elem) => {
        const $child = this.props.$(elem);
        children.push(
          <CheerioReactBind
            errorHandler={this.props.errorHandler}
            key={index}
            $={this.props.$}
            $elem={$child}
            tags={this.props.tags}
            tagRenderer={this.props.tagRenderer}
            location={`${
              this.props.location === "/" ? "" : this.props.location
            }/${$child.index()}:${$child.prop("tagName").toLowerCase()}`}
          />
        );
      });
    } else {
      children.push(this.props.$elem.text());
    }
    const state = {
      attributes: this.props.$elem.attr(),
      children
    };
    if (first) {
      this.state = state;
    } else {
      this.setState(state);
    }
  }

  render() {
    const { $elem } = this.props;
    const tagName = $elem.prop("tagName").toLowerCase();
    if (!this.props.tagRenderer && !this.props.tags) {
      throw new TypeError(
        'You must pass a "tagRenderer" prop or a "tags" prop.'
      );
    }

    if (this.props.tagRenderer) {
      const TagHander = this.props.tagRenderer;
      return (
        <TagHander
          location={this.props.location}
          tagName={tagName}
          attributes={this.state.attributes}
        >
          {this.state.children}
        </TagHander>
      );
    }

    if (!Object.prototype.hasOwnProperty.call(this.props.tags, tagName)) {
      const errMsg = `Unknown tag "${tagName}".`;
      if (this.props.errorHandler) {
        this.props.errorHandler(errMsg);
        return <div />;
      }
      throw new TypeError(errMsg);
    }
    const Component = this.props.tags[tagName];
    return (
      <Component {...this.state.attributes}>{this.state.children}</Component>
    );
  }
}

/* eslint-disable react/require-default-props, consistent-return */
CheerioReactBind.propTypes = {
  $: (props, propName) => {
    const $ = props[propName];
    if (!$) {
      return new Error('Prop type "$" is required for CheerioReactBind');
    }
    if (!$.root) {
      return new Error(
        'Prop type "$" must be a valid Cheerio object for CheerioReactBind'
      );
    }
  },
  $elem: (props, propName) => {
    const $elem = props[propName];
    if (!$elem) {
      return new Error('Prop type "$elem" is required for CheerioReactBind');
    }
    if (!$elem.data || !$elem.prop) {
      return new Error(
        'Prop type "$" must be a valid Cheerio element for CheerioReactBind'
      );
    }
  },
  // eslint-disable-next-line react/forbid-prop-types
  tags: PropTypes.object,
  tagRenderer: PropTypes.oneOfType([PropTypes.element, PropTypes.func]),
  errorHandler: PropTypes.func,
  location: PropTypes.string
};
/* eslint-enable react/require-default-props, consistent-return */

CheerioReactBind.defaultProps = {
  location: "/"
};
