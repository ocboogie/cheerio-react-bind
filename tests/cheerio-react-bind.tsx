import * as Cheerio from "cheerio";
import * as Enzyme from "enzyme";
import * as React from "react";

import CheerioReactBind, { TagRendererProps } from "../src/cheerio-react-bind";

const tags = {
  div: ({ children }) => <div>{children}</div>,
  h1: ({ children }) => <h1>{children}</h1>,
  p: ({ children }) => <p>{children}</p>,
  throw: ({ error }) => {
    if (error) {
      throw error;
    }
    return <div />;
  }
};

const $ = Cheerio.load(
  `
  <div foo=bar>
    <p>Paragraph</p>
  </div>
`,
  { xmlMode: true }
);

test("update the Cheerio dom when the React dom is updated", () => {
  const wrapper = Enzyme.mount(
    <CheerioReactBind tags={tags} $elem={$("div").first()} $={$} />
  );
  expect(wrapper.html()).toMatchSnapshot();

  $("div").append(`<div>test</div>`);
  $("div").update();
  expect(wrapper.html()).toMatchSnapshot();

  $("p").text("Paragraph test");
  $("p").update();
  expect(wrapper.html()).toMatchSnapshot();
});

test("able to have text nodes and tag nodes as children", () => {
  const $mock = Cheerio.load(
    `
    <div>
      foo
      <p>bar</p>
      baz
    </div>
  `,
    { xmlMode: true }
  );

  const wrapper = Enzyme.mount(
    <CheerioReactBind tags={tags} $elem={$mock("div").first()} $={$mock} />
  );
  expect(wrapper.html()).toMatchSnapshot();
});

describe("error handling", () => {
  (Error.prototype as any).suppressReactErrorLogging = true;
  test("throws when there is a unknown tag", () => {
    // This test will cause a lot of spam
    // Related issue https://github.com/airbnb/enzyme/issues/1255
    Enzyme.mount(
      <CheerioReactBind tags={tags} $elem={$("div").first()} $={$} />
    );
    $("div").append("<notag />");
    expect(() => {
      $("div").update();
    }).toThrowErrorMatchingSnapshot();
  });

  test("errorHandling runs when an error is thrown", () => {
    const mockErrorHandling = jest.fn();
    Enzyme.mount(
      <CheerioReactBind
        errorHandler={mockErrorHandling}
        tags={tags}
        $elem={$("div").first()}
        $={$}
      />
    );
    $("div").append("<notag />");
    $("div").update();

    expect(mockErrorHandling.mock.calls[0][0]).toBe(
      'There was an error rendering a tag at "/": "TypeError: Unknown tag name "notag"."'
    );
  });

  test("catches all errors", () => {
    const mockErrorHandling = jest.fn();
    const $mock = Cheerio.load(
      `
      <div>
        <throw error="oops" />
      </div>
    `,
      { xmlMode: true }
    );

    Enzyme.mount(
      <CheerioReactBind
        errorHandler={mockErrorHandling}
        tags={tags}
        $elem={$mock("div").first()}
        $={$mock}
      />
    );

    expect(mockErrorHandling.mock.calls[0][0]).toBe(
      'There was an error rendering a tag at "/0:throw/": "oops"'
    );
  });

  test("throws when neither tags nor tagRenderer are passed", () => {
    expect(() => {
      Enzyme.render(<CheerioReactBind $elem={$("div").first()} $={$} />);
    }).toThrow('You must pass a "tagRenderer" prop or a "tags" prop.');
  });

  test("use errorComponent when an error is thrown", () => {
    const mockErrorHandling = jest.fn();
    const $mock = Cheerio.load(
      `
      <div>
        <throw error="oops" />
      </div>
    `,
      { xmlMode: true }
    );

    const errorComponent = () => <h1>Oh no!</h1>;

    const wrapper = Enzyme.mount(
      <CheerioReactBind
        errorHandler={mockErrorHandling}
        errorComponent={errorComponent}
        tags={tags}
        $elem={$mock("div").first()}
        $={$mock}
      />
    );

    expect(wrapper.contains(<h1>Oh no!</h1>)).toBeTruthy();
  });

  (Error.prototype as any).suppressReactErrorLogging = false;
});

describe("tagRenderer", () => {
  test("gets passed children, attributes and tagName", () => {
    const $mock = Cheerio.load(
      `
    <foo bar="baz"><qux /></foo>
  `,
      { xmlMode: true }
    );
    const tagRenderer = jest.fn(() => <div />);
    Enzyme.mount(
      <CheerioReactBind
        tagRenderer={tagRenderer}
        $elem={$mock("foo").first()}
        $={$mock}
      />
    );
    expect(tagRenderer.mock.calls[0][0]).toMatchObject({
      attributes: { bar: "baz" },
      tagName: "foo"
    });
    expect(() =>
      Enzyme.shallow(tagRenderer.mock.calls[0][0].children[0])
    ).not.toThrow();
  });

  test("gets passed args from tagRendererArgs prop", () => {
    const tagRenderer = jest.fn(() => <div />);
    Enzyme.mount(
      <CheerioReactBind
        tagRenderer={tagRenderer}
        tagRendererArgs={{ foo: "bar" }}
        $elem={$("div").first()}
        $={$}
      />
    );
    expect(tagRenderer.mock.calls[0][0].foo).toBe("bar");
  });

  test("renders", () => {
    const $mock = Cheerio.load(
      `
    <foo bar="baz"><qux>quux</qux></foo>
  `,
      { xmlMode: true }
    );
    const tagRenderer: React.StatelessComponent<TagRendererProps> = ({
      tagName,
      children
    }) => {
      if (tagName === "foo") {
        return <span>{children}</span>;
      } else if (tagName === "qux") {
        return <h2>{children}</h2>;
      }
      return null;
    };
    const wrapper = Enzyme.mount(
      <CheerioReactBind
        tagRenderer={tagRenderer}
        $elem={$mock("foo").first()}
        $={$mock}
      />
    );
    expect(wrapper.html()).toMatchSnapshot();
  });
});

test("tagRenderer gets passed the location of the tag", () => {
  const $mock = Cheerio.load(
    `
    <div>
      <foo></foo>
      <bar><baz></baz></bar>
    </div>
`,
    { xmlMode: true }
  );
  const tagRenderer: React.StatelessComponent<TagRendererProps> = ({
    tagName,
    location,
    children
  }) => {
    if (tagName === "baz") {
      expect(location).toBe("/1:bar/0:baz/");
    }
    return <div>{children}</div>;
  };
  Enzyme.mount(
    <CheerioReactBind
      tagRenderer={tagRenderer}
      $elem={$mock("div").first()}
      $={$mock}
    />
  );
});
