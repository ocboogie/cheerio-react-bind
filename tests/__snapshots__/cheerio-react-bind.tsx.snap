// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`able to have text nodes and tag nodes as children 1`] = `
"<div>
      foo
      <p>bar</p>
      baz
    </div>"
`;

exports[`error handling throws when there is a unknown tag 1`] = `"There was an error rendering a tag at \\"/\\": \\"TypeError: Unknown tag name \\"notag\\".\\""`;

exports[`tagRenderer renders 1`] = `"<span><h2>quux</h2></span>"`;

exports[`update the Cheerio dom when the React dom is updated 1`] = `"<div><p>Paragraph</p></div>"`;

exports[`update the Cheerio dom when the React dom is updated 2`] = `"<div><p>Paragraph</p><div>test</div></div>"`;

exports[`update the Cheerio dom when the React dom is updated 3`] = `"<div><p>Paragraph</p><div>test</div></div>"`;
